build_daemon/lib/src/server.dart

```diff
@@ -0,0 +1,129 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:pool/pool.dart';
+import 'package:shelf/shelf_io.dart';
+import 'package:shelf_web_socket/shelf_web_socket.dart';
+import 'package:stream_transform/stream_transform.dart' hide concat;
+import 'package:watcher/watcher.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+import '../daemon_builder.dart';
+import '../data/build_options_request.dart';
+import '../data/build_request.dart';
+import '../data/build_target_request.dart';
+import '../data/log_to_paths_request.dart';
+import '../data/serializers.dart';
+import 'managers/build_target_manager.dart';
+import 'managers/channel_options_manager.dart';
+
+class Server {
+  final _isDoneCompleter = Completer();
+  final _buildTargetManager = BuildTargetManager();
+  final _buildOptionsManager = ChannelOptionsManager();
+  final _logPathsManager = ChannelOptionsManager();
+  final _pool = Pool(1);
+
+  HttpServer _server;
+  DaemonBuilder _builder;
+  final _channels = Set<WebSocketChannel>();
+  // Channels that are interested in the current build.
+  var _interestedChannels = Set<WebSocketChannel>();
+
+  final _subs = <StreamSubscription>[];
+
+  Server(this._builder, Stream<WatchEvent> changes) {
+    _forwardData();
+    _handleChanges(changes);
+  }
+
+  Future<void> get onDone => _isDoneCompleter.future;
+
+  Future<int> listen() async {
+    var handler = webSocketHandler((WebSocketChannel channel) async {
+      _channels.add(channel);
+
+      channel.stream.listen((message) async {
+        dynamic request;
+        try {
+          request = serializers.deserialize(jsonDecode(message as String));
+        } catch (_) {
+          print('Unable to parse message: $message');
```

Comment:
I wonder if printing is appropriate in the library code... Should we look at using a Logger instead?

Reply:
Yes. We should eventually make use of the logger package. I'll open an issue.